File: 152-maximum-product-subarray/maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int max_overall = nums[0];
        int max_curr = nums[0];
        int min_curr = nums[0];

        for (int i = 1; i < nums.size(); i++) {
            if (nums[i] < 0) {
                swap(max_curr, min_curr);
            }
            max_curr = max(nums[i], max_curr * nums[i]);
            min_curr = min(nums[i], min_curr * nums[i]);

            max_overall = max(max_overall, max_curr);
        }

        return max_overall;
    }
};

File: 152-maximum-product-subarray/maximum-product-subarray.go
func maxProduct(nums []int) int {
    max_product := math.MinInt
    curr_max, curr_min := 1, 1

    for i := range nums {
        if nums[i] < 0 {
            curr_max, curr_min = curr_min, curr_max
        }

        curr_max = max(curr_max * nums[i], nums[i])
        curr_min = min(curr_min * nums[i], nums[i])

        max_product = max(max_product, curr_max)
    }

    return max_product
}